### ISSUE TITLE
OpenSSL 1.1.1d - Cannot retrieve SNI from ClientHello

### ISSUE THREAD

**Issue by TheDechev** [openssl/openssl#10630]
On my running server (using **OpenSSL 1.1.1d**) I tried extracting the Server Name (SNI) extension from ClientHello messages using the following callback [API][1] by:

 - Calling `SSL_client_hello_get0_ext` with the `TLSEXT_TYPE_server_name` type as done in the [repo][2].
 - Calling `SSL_get_servername` with `TLSEXT_NAMETYPE_host_name` [as documented][3].

Yet nothing seems to work, the SNI is either NULL or garbage.

One thing that I noticed was that only after finishing the ClientHello callback I was able to extract it, but at that point it's too late.
Has someone encountered a similar issue? Am I doing something wrong?

  [1]: https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_client_hello_cb.html
  [2]: https://github.com/openssl/openssl/blob/12a765a5235f181c2f4992b615eb5f892c368e88/test/handshake_helper.c#L162
  [3]: https://www.openssl.org/docs/man1.1.1/man3/SSL_get_servername.html

**Comment by kaduk** [comment]
From which application callback(s) did you try those two methods?

**Comment by TheDechev** [comment]
@kaduk - Both methods were called from the invoked client hello callback.
As someone cleared to me the `SSL_get_servername` should be invoked directly from the server name callback instead of the client hello callback itself. (Which didn't solve my issue)

BUT, with further digging we managed to find a way to retrieve the extension in the client hello callback by parsing the extension manually. (as done [here](https://github.com/openssl/openssl/blob/12a765a5235f181c2f4992b615eb5f892c368e88/test/handshake_helper.c#L162)).

I was hoping for a direct API call for the string but that's close enough.

**Comment by mattcaswell** [comment]
The ClientHello callback happens very early in the process - before extensions have been parsed. Therefore API calls like `SSL_get_servername` do not work there. You have to parse the extensions directly yourself.

**Comment by kaduk** [comment]
SNI in particular has a very complicated structure (and yet is commonly used) so even from the start I had considered adding a dedicated API for it.  But that was not part of the core functionality, and so I left it out of the initial commit.  Perhaps there is enough demand to introduce it.

**Comment by richsalz** [comment]
@kaduk, do you know of any people who are putting anything other than a NameType of "hostname" there?  RFC 6066 went overboard IMO.

**Comment by kaduk** [comment]
@richsalz no; the TLS WG has effectively admitted that the name type joint has "rusted shut" and we only ever see "hostname"s, which is most of why I say the SNI structure is complicated.

**Comment by TheDechev** [comment]
@kaduk In my opinion it would be really helpful to have dedicated API for it :)

**Comment by magsys** [comment]
@kaduk Another vote for SSL_client_hello_get0_servername, without it replacing the  ServerName Callback is not a five minute job, but a research project into the construction of extensions.
Ditto  SSL_client_hello_get0_alpn since that is extension I need first.

**Comment by kaduk** [comment]
Thanks for chiming in!
In the interim, I'll note (mostly for others who search and run into this post) that the problem is to some extent already solved for the test suite; see https://github.com/openssl/openssl/blob/8658feddea6aef5cf5cbb1cfbf6b1817fa432051/test/handshake_helper.c#L158-L187

**Comment by magsys** [comment]
Thanks, this new callback has finally allowed me to make the Let's Encrypt tls-alpn-01 challenge work properly, it needed a new certificate to be loaded into the context for that ALPN, but by the time the ALPN callback is made it's too late and the new certificate was ignored,  Now I can select a different context instead in ClientHello.  Still need to use the ALPN callback since that is the only way to send the ALPN back to the client.  Spent days fighting this last summer. 

Strangely, I just noticed our SSL library uses SSL_set_tlsext_debug_callback which appears to one of many undocumented functions, and I could have got the ALPN from there if I'd known more about extensions.
